File: js/stack/evaluate-polish-notation.js
/**
 * https://leetcode.com/problems/evaluate-reverse-polish-notation
 * 
 * @param {string[]} tokens
 * @return {number}
 */
const evalRPN = (tokens) => {
    const stack = [];
    const operators = ['+', '-', '*', '/'];


    for (let i = 0; i < tokens.length; i++) {
        if (!operators.includes(tokens[i])) {
            stack.push(Number(tokens[i]));
        } else {
            const second = stack.pop();
            const first = stack.pop();

            if (tokens[i] === "+") { 
                stack.push(first + second);
            }

            if (tokens[i] === "-") { 
                stack.push(first - second);
            }

            if (tokens[i] === "*") { 
                stack.push(first * second);
            }

            if (tokens[i] === "/") { 
                stack.push(Math.trunc(first / second));
            }
        }
    };

    return stack;
}

File: python/stack/evaluate-polish-notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []

        for t in tokens:
            if t == '+':
                stack.append(stack.pop() + stack.pop())
            elif t == '-':
                first, second = stack.pop(), stack.pop()
                stack.append(second - first)
            elif t == '/':
                first, second = stack.pop(), stack.pop()
                stack.append(math.trunc(second / first))
            elif t == '*':
                stack.append(stack.pop() * stack.pop())
            else:
                stack.append(int(t))

        return stack[0]
        

# https://leetcode.com/problems/evaluate-reverse-polish-notation/
